design-proposals/cpu-hotplug.md

```diff
@@ -0,0 +1,146 @@
+# Overview
+This proposal is about adding CPU hotplug support to KubeVirt.
+CPU hotplug will allow users to dynamically add/remove CPU resources to/from running VMs.
+While this design is focused on CPU hotplug, it aims to provide a generic API design for resource hotplug that will be reusable for things like memory.
+
+## Motivation
+CPU hotplug is now a common virtualization feature, but KubeVirt lacks support for it.
+Live vertical scaling of VM resources in general can be very useful in a variety of scenarios.
+
+## Goals
+- Hotplug and hot-unplug of CPU resources from VMs should be achievable by editing the VM object itself (no sub-resources)
+- Underlying technology used for hotplug should be independent of the VM API, meaning we should be able to swap out how hotplug is technically performed in the future without impacting API compatibility
+- Implementation should be achievable today, with Kubernetes APIs that are at least in beta. Unfortunately, at the time of writing, the Kubernetes vertical pod scaling API is only alpha
+
+## Non Goals
+- Revising existing hotplug mechanisms like Volume hotplug
+- Dynamically applying affinity rules to VMIs
+
+## Definition of Users
+A user with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+## User Stories
+As a VM user, I would like to dynamically increase or decrease the number of CPUs used by a VM without requiring the VM to restart.
+
+## Repos
+kubevirt/kubevirt
+
+# Design
+In VM objects, the existing CPU `sockets` field will now be dynamic, provided that it has been declared as such.
+Declaring the field as dynamic will be done by adding a `cpu` entry to a new section under the VM spec, called `liveUpdateFeatures`.
+Under the `cpu` entry, it will be possible to define a maximum number of sockets. That number is needed by LibVirt and will default to 4 times the initial number of sockets if not set.
+
+Increasing the number of CPU sockets will not only add CPUs to the guest but also increase the CPU resources available to the virt-launcher pod, via a migration (more on that in the Implementation section).
+The way LibVirt handles CPU hotplug is by exposing the maximum number of CPUs to the guest and turning off the unused CPUs. In this document, when we refer to "adding" CPUs to the guest, we're really just turning them on. This is why we need to define a maximum number of sockets in advance.
+
+## Drawbacks / Limitations
+- Since CPU hot-(un)plug involves a live migration, VMs will have to be `LiveMigratable` to enable CPU hotplug
+- Each disabled vCPU, so `(maxSockets - sockets) * cores * thread`, consumes 8MiB of overhead memory
+- This feature will be incompatible with CPU requests/limits, at least initially, since those values need to dynamically scale according to the current number of enabled CPUs
```

Comment:
Won't the target pod be created with the updated topology and hence increase the requested overhead?

Reply:
```suggestion
- This feature makes CPU requests/limits second-class entries, as they would be automatically recalculated by KubeVirt based on the number of sockets.
```
+1 for a rephrase ^^^

Reply:
@dankenigsberg You stated that as a user you want the CPU hotplug to happen automatically, w/o additional API fields like `liveUpdateFeatures`, but how will it fit with this drawback? and if the suggestion is to gate CPU hotplug in KV CR, what about multi-tenant envs?

Reply:
Hmm. I did not think about this issue. I cannot imagine a user changing the number of cores but wanting it to take effect only in the far future. However I can image someone being surprised by by the request/limits suddenly recalculated.

I think that the way to solve this is to have KubeVirt recalculate the requests/limits of the virt-launcher pod and only if requests/limits were not specified. The VM spec is the input to KubeVirt, controlled by humans. It would be a violation of layering if KubeVirt is to edit its own spec.

I don't think that it is important to block hot-plug on the cluster level and certainly not on the tenant level. VM owners can decide to use it only when they need it. I think that `liveUpdateFeatures` introduces complexity that is not really needed by users; it is an expensive commitment that clutters the API and forces us to add many tests. If we believe that we must cater to users who do not want hot-plug, a cluster-wide feature gate would be less expensive.